File: server/app/routers/uploads.py
from __future__ import annotations
"""POST /api/v1/upload"""
import os
import uuid

from fastapi import APIRouter, Depends, UploadFile, File, Form
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import LogRecord, Upgrade, Box
from app.services.socketio_handler import emit_to_box
from app.utils import version_to_sort

router = APIRouter(prefix="/api/v1/upload", tags=["uploads"])


@router.post("")
async def upload(type: str = "", no: str = "", model: str = "", appVersion: str = "",
                 file: UploadFile = File(...), db: AsyncSession = Depends(get_db)):
    os.makedirs("uploads", exist_ok=True)
    ext = os.path.splitext(file.filename or "")[1]
    filename = f"{uuid.uuid4().hex}{ext}"
    file_path = os.path.join("uploads", filename)
    content = await file.read()
    with open(file_path, "wb") as f:
        f.write(content)

    if type == "logfile":
        log = LogRecord(no=no, originname=file.filename or "", name=filename, path=file_path)
        db.add(log)
        await db.commit()
    elif type == "upgrade":
        from app.utils import md5 as calc_md5
        up = Upgrade(
            model=model, appVersion=appVersion, sortAppVersion=version_to_sort(appVersion),
            originname=file.filename or "", name=filename, path=os.path.join("uploads", filename),
            md5=calc_md5(content),
        )
        db.add(up)
        await db.commit()
        result = await db.execute(select(Box).where(Box.status == "正常"))
        for box in result.scalars().all():
            await emit_to_box(box.no, {"type": "COMMAND", "content": {"cmd": "upgrade"}})

    return {"errcode": 0}

File: smarttriage/lib/controllers/uploads.js
'use strict';
const path = require('path');
const Utils = require('../utils');
const Model = require('../models');
const Log = Model('Log');
const Upgrade = Model('Upgrade');
const logger = Utils.logger(__filename);
const readFileMd5 = Utils.crypto.readFileMd5;
const upgrades = require('./upgrades');

module.exports = {
  async upload(ctx) {
    try {
      const query = ctx.request.query;
      const params = ctx.request.body;
      const type = query.type;
      const filePath = path.resolve(__dirname, `../../${ctx.request.files.file.path}`);
      const name = ctx.request.files.file.path.split('\\')[ctx.request.files.file.path.split('\\').length - 1];
      if (type == 'logfile') {
        const log = new Log({
          no: params.no,
          originname: ctx.request.files.file.name,
          name,
          path: filePath,
          ct: new Date(),
          ut: new Date()
        });
        await log.save();
      } else if (type == 'upgrade') {
        const versions = params.appVersion.split('.');
        const sortAppVersion = Number(`${10000 + Number(versions[0])}${10000 + Number(versions[1])}${10000 + Number(versions[2])}`);
        const md5 = await readFileMd5(filePath);
        const upgrade = new Upgrade({
          model: params.model,
          appVersion: params.appVersion,
          sortAppVersion,
          originname: ctx.request.files.file.name,
          name,
          path: filePath,
          md5,
          ct: new Date(),
          ut: new Date()
        });
        await upgrade.save();
        await upgrades.sendUpgradeCmd(ctx);
      }

      ctx.body = {
        errcode: 0
      };
    } catch (e) {
      logger.error('error', e);
      ctx.status = 500;
      ctx.body = {
        errcode: 10000,
        errmsg: 'server error'
      };
    }
  }
};
